addons/account/models/account_bank_statement.py

```diff
@@ -693,17 +679,63 @@ def _prepare_move_line_default_vals(self, counterpart_account_id=None):
             'account_id': counterpart_account_id,
             'move_id': self.move_id.id,
             'partner_id': self.partner_id.id,
-            'currency_id': foreign_currency.id,
+            'currency_id': currencies['foreign'].id,
             'amount_currency': -transaction_amount,
             'debit': -company_amount if company_amount < 0.0 else 0.0,
             'credit': company_amount if company_amount > 0.0 else 0.0,
         }
         return [liquidity_line_vals, counterpart_line_vals]
 
+    def _get_currencies(self):
+        company_currency = self.journal_id.company_id.currency_id
+        journal_currency = self.journal_id.currency_id or company_currency
+        foreign_currency = self.foreign_currency_id or journal_currency or company_currency
+        return {
+            'company': company_currency,
+            'journal': journal_currency,
+            'foreign': foreign_currency,
+        }
+
+    def _get_amounts(self):
+        currencies = self._get_currencies()
+        journal_amount = self.amount
+        if currencies['foreign'] == currencies['journal']:
+            transaction_amount = journal_amount
+        else:
+            transaction_amount = self.amount_currency
+        if currencies['journal'] == currencies['company']:
+            company_amount = journal_amount
+        elif currencies['foreign'] == currencies['company']:
+            company_amount = transaction_amount
+        else:
+            company_amount = currencies['journal']._convert(journal_amount, currencies['company'], self.journal_id.company_id, self.date)
+        return company_amount, journal_amount, transaction_amount
+
     # -------------------------------------------------------------------------
     # COMPUTE METHODS
     # -------------------------------------------------------------------------
 
+    def _compute_cumulated_balance(self):
+        for st_line in self:
+            st_line.cumulated_balance = st_line.move_id.line_ids.filtered(
+                lambda aml: aml.account_id == st_line.journal_id.default_account_id).cumulated_balance
+
+    @api.depends('journal_id')
+    def _compute_statement_id(self):
+        for line in self.filtered(lambda l: l.statement_id.journal_id != l.journal_id):
```

Comment:
When the user changes the journal in the form view (it is editable when the journal is not in the context), the statement should be recalculated, but I'll check it again, it doesn't look right. The goal was if the user changes the journal but has selected the statement for new journal before. It is not possible now because of the domain on the statement.